### PR TITLE
CASSANDRA-19980: Remove SparkSQL dependency from CassandraBridge so t…

### PR DESCRIPTION
**Pull request by jberragan**
…hat it can be used independent from Spark

### CODE REVIEW

**Comment by jberragan** [comment]
Green CircleCI: https://app.circleci.com/pipelines/github/jberragan/cassandra-analytics/117/workflows/248920f4-b236-4199-a3c9-68f68a8f5693

**Comment by jberragan** [comment]
Latest green CI: https://app.circleci.com/pipelines/github/jberragan/cassandra-analytics/118/workflows/1dc1d879-865f-44b3-8eec-7363529bf51f

File: cassandra-bridge/build.gradle [hunk 1]
```diff
@@ -40,14 +40,32 @@ publishing {
...
+configurations {
+    testArtifacts
+}
+task testJar(type: Jar) {
+    archiveBaseName = "${project.name}-bridge-test"
+    from sourceSets.test.output
+}
+artifacts {
+    testArtifacts testJar
+}
```
Comment (yifan-c) [inline cassandra-bridge/build.gradle]: Using test-fixture instead? https://docs.gradle.org/current/userguide/java_testing.html#sec:java_test_fixtures

File: cassandra-bridge/build.gradle [hunk 2]
```diff
@@ -40,14 +40,32 @@ publishing {
...
-    api(project(':cassandra-analytics-spark-converter'))
-    compileOnly(group: "${sparkGroupId}", name: "spark-core_${scalaMajorVersion}", version: "${project.rootProject.sparkVersion}")
-    compileOnly(group: "${sparkGroupId}", name: "spark-sql_${scalaMajorVersion}", version: "${project.rootProject.sparkVersion}")
 
+    implementation "org.slf4j:slf4j-api:${slf4jApiVersion}"
+    compileOnly "com.esotericsoftware:kryo-shaded:${kryoVersion}"
+    compileOnly "com.google.guava:guava:${guavaVersion}"
+    compileOnly "com.fasterxml.jackson.core:jackson-annotations:${jacksonVersion}"
+    compileOnly "com.fasterxml.jackson.core:jackson-core:${jacksonVersion}"
+    compileOnly "com.fasterxml.jackson.core:jackson-databind:${jacksonVersion}"
```
Comment (yifan-c) [inline cassandra-bridge/build.gradle]: nit: use the same parenthesis style as the other lines. `compileOnly(...)`

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java [hunk 3]
```diff
@@ -33,12 +33,15 @@
...
 import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.tuple.Pair;
 
+import org.apache.cassandra.spark.data.converter.SparkSqlTypeConverter;
 import org.jetbrains.annotations.NotNull;
 
 public final class CassandraBridgeFactory
 {
-    private static final Map<String, CassandraBridge> CASSANDRA_BRIDGES = new ConcurrentHashMap<>(CassandraVersion.values().length);
+    private static final Map<String, Pair<CassandraBridge, SparkSqlTypeConverter>> CASSANDRA_BRIDGES =
```
Comment (yifan-c) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java]: nit: add a comment to indicate what is the key String value represents

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java [hunk 4]
```diff
@@ -75,12 +78,32 @@ public static CassandraBridge get(@NotNull CassandraVersionFeatures features)
         return get(getCassandraVersion(features));
     }
 
+    @NotNull
+    public static SparkSqlTypeConverter getSparkSql(@NotNull CassandraVersionFeatures features)
+    {
+        return getSparkSql(getCassandraVersion(features));
+    }
+
```
Comment (yifan-c) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java]: Please locate the method together with the other `getSparkSql` overloads.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java [hunk 5]
```diff
@@ -155,6 +157,11 @@ public CassandraVersion version()
      */
     public abstract CassandraBridge bridge();
 
+    public SparkSqlTypeConverter typeConverter()
```
Comment (yifan-c) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java:163]: Add java doc

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java [hunk 6]
```diff
@@ -130,14 +153,20 @@ private static String typesResourceName(@NotNull String label)
...
     private static String jarResourceName(String... parts)
     {
         return "/bridges/" + String.join("-", parts) + ".jar";
     }
 
     @NotNull
     @SuppressWarnings("unchecked")
-    private static CassandraBridge create(@NotNull String label)
+    private static Pair<CassandraBridge, SparkSqlTypeConverter> create(@NotNull String label)
```
Comment (yifan-c) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java]: The method signature looks a bit odd. `CassandraBridgeFactory#create` returns a pair.. 

How about create a dedicated data class, say `VersionedCassandraBridge`, that wraps both `CassandraBridge` and `SparkSqlTypeConverter`?  Then `create` and `get` methods return the `VersionedCassandraBridge`. `getSparkSql` may be removed. 

```java
public class VersionedCassandraBridge
{
    public final CassandraBridge cassandraBridge;
    public final SparkSqlTypeConverter sparkSqlTypeConverter;
}
```

**Review by yifan-c — COMMENTED** [review]
Looks good in general. I have some cosmetic comments. 

I also verified that there is no more spark in the bridge runtime dependencies.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java (hunk 6 above)
Comment (jberragan) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java]: Ok, it seems a bit OTT given it is private/internal to `CassandraBridgeFactory`

File: cassandra-bridge/build.gradle (hunk 1 above)
Comment (jberragan) [reply cassandra-bridge/build.gradle]: It appears `java-test-fixtures` does not support JDK8

Comment (jberragan) [reply cassandra-bridge/build.gradle]: I reverted to using `testArtifacts`

**Comment by jberragan** [comment]
Latest green CircleCI: https://app.circleci.com/pipelines/github/jberragan/cassandra-analytics/121/workflows/f2cf72cb-dd56-48dc-9c1d-06965b21d864

File: cassandra-bridge/build.gradle (hunk 1 above)
Comment (yifan-c) [reply cassandra-bridge/build.gradle]: Sure. thanks for trying it out. Sadly, such limitation was not mentioned in the plugin page.

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java (hunk 6 above)
Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/bridge/CassandraBridgeFactory.java]: yeah. I find a proper data class is more readable and maintainable than `Pair` in the long run.

**Review by frankgh — COMMENTED** [review]
Looks good in general. I suggested a couple of change offline regarding test fixtures and moving the spark package name to analytics in the bridge subproject

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java [hunk 7]
```diff
@@ -46,7 +46,7 @@
 import org.apache.cassandra.spark.sparksql.filters.PartitionKeyFilter;
 import org.apache.cassandra.spark.sparksql.filters.PruneColumnFilter;
 import org.apache.cassandra.spark.sparksql.filters.SparkRangeFilter;
-import org.apache.cassandra.spark.stats.Stats;
+import org.apache.cassandra.analytics.stats.Stats;
```
Comment (yifan-c) [inline cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java:49]: please rename package to `bridge` instead.

**Review by yifan-c — COMMENTED** [review]
Can package `org.apache.cassandra.spark.reader` rename to `org.apache.cassandra.bridge.reader` too?

File: cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java (hunk 7 above)
Comment (jberragan) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java:49]: I don't think that is really in scope for this change. I don't want PRs to become a bucket for every change everyone wants to squeeze in. I'm tempted to revert https://github.com/apache/cassandra-analytics/pull/88/commits/67590a268f095ef6f91bd7e1d195300cd75dac71 as well.

Comment (jberragan) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java:49]: There are plenty of util and standalone classes in `org.apache.cassandra.spark.reader` that could be used downstream, I'm not saying we can't relocate especially for major version change but it seems worth a standalone change.

Comment (yifan-c) [reply cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/data/DataLayer.java:49]: Reverting and relocating in a separate patch sounds good to me.

**Pull request merged**